halo2-ecc/src/bn254/bls_signature.rs

```diff
@@ -41,55 +41,17 @@ impl<'chip, F: PrimeField> BlsSignatureChip<'chip, F> {
-        // Checking element from G1 is on curve also check that it's in subgroup G1 since G1 has cofactor of 1
-        g1_chip.assert_is_on_curve::<halo2_base::halo2_proofs::halo2curves::bn256::G1Affine>(
-            ctx,
-            &g1_assigned,
-        );
-
-        let hash_m_assigned = self.pairing_chip.load_private_g2_unchecked(ctx, msghash);
-        g2_chip.assert_is_on_curve::<halo2_base::halo2_proofs::halo2curves::bn256::G2Affine>(
-            ctx,
-            &hash_m_assigned,
-        );
+        let g1_assigned = self.pairing_chip.load_private_g1(ctx, g1);
 
-        let mut signature_agg_assigned =
-            self.pairing_chip.load_private_g2_unchecked(ctx, signatures[0]);
-        g2_chip.field_chip.enforce_less_than(ctx, signature_agg_assigned.x().clone());
-        let mut pubkey_agg_assigned = self.pairing_chip.load_private_g1_unchecked(ctx, pubkeys[0]);
-        g1_chip.field_chip.enforce_less_than(ctx, pubkey_agg_assigned.x().clone());
+        let hash_m_assigned = self.pairing_chip.load_private_g2(ctx, msghash);
 
-        // loop through signatures and aggregate them
-        for (index, signature) in signatures.iter().enumerate() {
-            if index > 0 {
-                let signature_assigned =
-                    self.pairing_chip.load_private_g2_unchecked(ctx, *signature);
-                g2_chip.field_chip.enforce_less_than(ctx, signature_assigned.x().clone());
-                g2_chip
-                    .assert_is_on_curve::<halo2_base::halo2_proofs::halo2curves::bn256::G2Affine>(
-                        ctx,
-                        &signature_assigned,
-                    );
-                signature_agg_assigned =
-                    g2_chip.add_unequal(ctx, &signature_agg_assigned, &signature_assigned, false);
-            }
-        }
+        let signature_points =
+            signatures.iter().map(|pt| g2_chip.assign_point(ctx, *pt)).collect::<Vec<_>>();
+        let signature_agg_assigned = g2_chip.sum::<G2Affine>(ctx, signature_points);
 
-        // loop through pubkeys and aggregate them
-        for (index, pubkey) in pubkeys.iter().enumerate() {
-            if index > 0 {
-                let pubkey_assigned = self.pairing_chip.load_private_g1_unchecked(ctx, *pubkey);
-                g1_chip.field_chip.enforce_less_than(ctx, pubkey_assigned.x().clone());
-                g1_chip
-                    .assert_is_on_curve::<halo2_base::halo2_proofs::halo2curves::bn256::G1Affine>(
-                        ctx,
-                        &pubkey_assigned,
-                    );
-                pubkey_agg_assigned =
-                    g1_chip.add_unequal(ctx, &pubkey_agg_assigned, &pubkey_assigned, false);
-            }
-        }
+        let pubkey_points =
+            pubkeys.iter().map(|pt| g1_chip.assign_point(ctx, *pt)).collect::<Vec<_>>();
+        let pubkey_agg_assigned = g1_chip.sum::<G1Affine>(ctx, pubkey_points);
 
```

Comment:
ah I also missed last time -- I think you want the result to just be a boolean 0 or 1 for whether the signature verified. Right now it's returning a `FqPoint`. So you should add the check that `result == Fq12(1)`

Reply:
I ha errors using `result == Fq12(1) / Fq12::one()` directly, but I assume you meant to compare the `format()` of the 2 parameters instead

Reply:
I should be more specific. I meant you should return an `AssignedValue<F>` that is constrained to be either `0` or `1`, and this value should output 1 if and only if the BLS signature is verified successfully.

Right now your `result` is an `Fp12Point`. You need to do (syntax might need to be fixed):
```
let fp12_one = fp12_chip.load_constant(ctx, Fq12::one());
fp12_chip.is_equal(ctx, result, fp12_one)
```

Reply:
Got it!

In the test `fn bls_signature_test()`, I have a not so pretty way of comparing `result: AssignedValue<F>` with `actual_result: Gt`. I'm not sure if there's a better way?